analyzers/src/SonarAnalyzer.CSharp/Rules/UnusedPrivateMember.cs

```diff
@@ -129,6 +91,58 @@ public sealed class UnusedPrivateMember : SonarDiagnosticAnalyzer
                         });
                 });
 
+        private static void NamedSymbolAction(SymbolAnalysisContext context, ConcurrentBag<ISymbol> removableInternalTypes)
+        {
+            var namedType = (INamedTypeSymbol)context.Symbol;
+            var privateSymbols = new HashSet<ISymbol>();
+            var fieldLikeSymbols = new BidirectionalDictionary<ISymbol, SyntaxNode>();
+            if (GatherSymbols((INamedTypeSymbol)context.Symbol, context.Compilation, privateSymbols, removableInternalTypes, fieldLikeSymbols)
+                && privateSymbols.Any()
+                && new CSharpSymbolUsageCollector(context.Compilation, privateSymbols) is var usageCollector
+                && VisitDeclaringReferences(namedType, usageCollector, context.Compilation, includeGeneratedFile: true))
+            {
+                var diagnostics = GetDiagnosticsForUnusedPrivateMembers(usageCollector, privateSymbols, Private, fieldLikeSymbols)
+                                        .Concat(GetDiagnosticsForUsedButUnreadFields(usageCollector, privateSymbols));
+                foreach (var diagnostic in diagnostics)
+                {
+                    context.ReportDiagnosticIfNonGenerated(diagnostic);
+                }
+            }
+        }
+
+        private static bool GatherSymbols(INamedTypeSymbol namedType,
+                                          Compilation compilation,
+                                          HashSet<ISymbol> privateSymbols,
+                                          ConcurrentBag<ISymbol> internalSymbols,
+                                          BidirectionalDictionary<ISymbol, SyntaxNode> fieldLikeSymbols)
+        {
+            if (namedType.ContainingType != null
+                // We skip top level statements since they cannot have fields. Other declared types are analyzed separately.
+                || namedType.IsTopLevelProgram()
+                || namedType.DerivesFromAny(IgnoredTypes)
+                // Collect symbols of private members that could potentially be removed
+                || RetrieveRemovableSymbols(namedType, compilation) is not { } removableSymbolsCollector)
+            {
+                return false;
+            }
+
+            CopyRetrievedSymbols(removableSymbolsCollector, privateSymbols, internalSymbols, fieldLikeSymbols);
+
+            // Collect symbols of private members that could potentially be removed for the nested classes
+            foreach (var syntaxReference in namedType.DeclaringSyntaxReferences.Where(r => !r.SyntaxTree.IsGenerated(CSharpGeneratedCodeRecognizer.Instance, compilation)))
+            {
+                foreach (var declaration in syntaxReference.GetSyntax().ChildNodes().OfType<BaseTypeDeclarationSyntax>().ToList())
+                {
```

Comment:
We can reduce a bit the imbrication by using:
```
foreach (var declaration in namedType.DeclaringSyntaxReferences
                                                 .Where(r => !r.SyntaxTree.IsGenerated(CSharpGeneratedCodeRecognizer.Instance, compilation))
                                                 .SelectMany(x => x.GetSyntax().ChildNodes().OfType<BaseTypeDeclarationSyntax>()))
            {
```

See also the code contained an extra `.ToList()` which was not needed (and was probably costly in terms of perf)